event/listener.php

```diff
@@ -18,14 +21,147 @@
 class listener implements EventSubscriberInterface
 {
 	/**
-	 * Assign functions defined in this class to event listeners in the core
+	 * @var manager Topic prefixes manager
+	 */
+	protected $manager;
+
+	/**
+	 * @var request Request object
+	 */
+	protected $request;
+
+	/**
+	 * @var user User object
+	 */
+	protected $user;
+
+	/**
+	 * @inheritdoc
+	 */
+	public static function getSubscribedEvents()
+	{
+		return [
+			'core.posting_modify_template_vars'			=> 'add_to_posting_form',
+			'core.posting_modify_submit_post_before'	=> 'submit_prefix_data',
+			'core.submit_post_modify_sql_data'			=> 'save_prefix_to_topic',
+		];
+	}
+
+	/**
+	 * Listener constructor
+	 *
+	 * @param manager $manager Topic prefixes manager
+	 * @param request $request Request object
+	 * @param user    $user    Language object
+	 */
+	public function __construct(manager $manager, request $request, user $user)
+	{
+		$this->manager = $manager;
+		$this->request = $request;
+		$this->user = $user;
+	}
+
+	/**
+	 * Update the posting page form with template vars
+	 * for the topic prefix drop-down menu.
+	 *
+	 * @param \phpbb\event\data $event Event data object
+	 * @return null
+	 */
+	public function add_to_posting_form($event)
+	{
+		if (!$this->is_new_topic($event))
+		{
+			return;
+		}
+
+		$this->user->add_lang_ext('phpbb/topicprefixes', 'topic_prefixes');
+
+		// Get prefixes for the current forum
+		$prefixes = $this->manager->get_active_prefixes($event['forum_id']);
+
+		// Get the current prefix selected
+		$selected = $this->get_selected_prefix($event);
+
+		$event['page_data'] = array_merge($event['page_data'], [
+			'PREFIXES'			=> $prefixes,
+			'SELECTED_PREFIX'	=> array_key_exists($selected, $prefixes) ? $prefixes[$selected]['prefix_tag'] : '',
+		]);
+	}
+
+	/**
+	 * Prepare topic prefix data for post submission
+	 *
+	 * @param \phpbb\event\data $event Event data object
+	 * @return null
+	 */
+	public function submit_prefix_data($event)
+	{
+		if (!($selected = $this->request->variable('topic_prefix', 0)))
```

Comment:
I don't see any reason to, I kind of like it (i like it when the first bit of code is a return condition)